File: FWI/Python/tutorial/tutorial_hourly_FWI.py
#### Hourly Fire Weather Index (FWI) Tutorial ####
# April 2025 (last updated February 2026)
#
# This script was designed to go with a tutorial on the NG-CFFDRS website to inform
# users how to use scripts associated with FWI2025. Follow along with the
# 'Hourly FWI Tutorial - Python' page on the NG-CFFDRS website:
# https://cffdrs.github.io/website_en/tutorials/Hourly_FWI_Python
# This tutorial will demonstrate how to generate FWI2025 outputs based on an
# input from a CSV file. The method will differ if using another source file type or
# if integrating code into existing fire management systems. This tutorial
# assumes the user has a working level knowledge of Python.
##############################################################################

### Load packages ###
# Run `pip install` to install any you are missing.
import pandas as pd
import sys

### Load functions and data ###
# This tutorial will refer to file locations as structured in the GitHub repository.
# Specify the local path to the cffdrs-ng folder. It can accept absolute paths (e.g.
# starting with "C:/") or relative paths (e.g. "./" if the working directory is
# cffdrs-ng). Change the path strings to match your file layout if it is different.
path_prefix = "CHANGE/TO/PATH/TO/cffdrs-ng/"
sys.path.append(path_prefix + "FWI/Python")

# Load the files containing the variables and functions to calculate FWI2025.
from NG_FWI import hFWI
from daily_summaries import generate_daily_summaries

# Load the input weather station data file.
data = pd.read_csv(path_prefix + "data/PRF2007_hourly_wx.csv")

# Print the column names, data should contain 12 columns
print(data.columns)
# Index(['id', 'lat', 'long', 'timezone', 'yr', 'mon', 'day', 'hr', 'temp', 'rh', 
#        'ws', 'prec'],
#       dtype='object')

### Run FWI2025 ###
# hFWI() is the function that calculates hourly FWI codes in FWI2025. It can
# handle multiple stations and years/fire seasons (not shown in this tutorial).
# For details you can run the help.
help(hFWI)
# Help on function hFWI in module NG_FWI:

# hFWI(
#     df_wx,
#     timezone=None,
#     ffmc_old=85.0,
#     mcffmc_old=None,
#     dmc_old=6.0,
#     dc_old=15.0,
#     mcgfmc_matted_old=16.3075131042516,
#     mcgfmc_standing_old=16.3075131042516,
#     prec_cumulative=0.0,
#     canopy_drying=0,
#     silent=False,
#     round_out=4
# )
#     ##
#     # Calculate hourly FWI indices from hourly weather stream.
#     #
#     # @param    df_wx               hourly values weather stream
# ...

# For this tutorial, we will leave all the optional parameters to default.
data_fwi = hFWI(data)
# ########
# FWI2025 (YYYY-MM-DD)

# Startup values used:
# FFMC = 85.0 % and mcffmc = None
# DMC = 6.0 and DC = 15.0
# mcgfmc matted = 16.3075 % and standing = 16.3075 %
# cumulative precipitation = 0.0 mm and canopy drying = 0

# Running PRF for 2007
# ########

# Output is a DataFrame, with FWI calculations appended after the input columns.
# Save the output as a CSV file (overrides any preexisting file).
data_fwi.to_csv(path_prefix + "PRF2007_hourly_FWI.csv", index = False)

# Print the last two rows of the standard moisture codes and fire behaviour indices.
standard_components = ['ffmc', 'dmc', 'dc', 'isi', 'bui', 'fwi']
print(data_fwi.loc[:, standard_components].tail(2))
#          ffmc     dmc        dc     isi     bui     fwi
# 2621  80.2669  3.0924  219.5169  1.6423  5.9745  0.7650
# 2622  82.1759  3.3503  220.0533  2.1462  6.4550  1.2023

# Print a simple summary of the standard FWI components.
print(data_fwi.loc[:, standard_components].describe())
#               ffmc          dmc           dc          isi          bui          fwi
# count  2623.000000  2623.000000  2623.000000  2623.000000  2623.000000  2623.000000
# mean     70.622770    17.487231   136.122169     2.158163    23.971546     3.877128
# std      22.973326    13.331302    76.454242     2.233486    17.499332     4.438263
# min       0.000000     0.000000    15.409200     0.000000     0.000000     0.000000
# 25%      67.866500     5.924600    72.276400     0.787700     8.407750     0.422050
# 50%      79.377900    16.460900   133.062700     1.395800    22.623400     2.407500
# 75%      85.177050    26.511200   188.638500     2.978500    34.605350     5.929900
# max      94.397300    50.174000   292.950100    17.595800    66.148100    25.135200

# Compare your outputs with our standard outputs in PRF2007_standard_hourly_FWI.csv

### Calculate daily summaries ###
# Calculate outputs like peak burn time and number of hours of spread potential.
report = generate_daily_summaries(data_fwi)
# ########
# FWI2025: Daily Summaries (YYYY-MM-DD)

# Summarizing PRF to daily
# ########

# Print a simple summary of the daily report.
daily_components = ["peak_hr", "duration", "isi_smooth", "dsr"]
print(report[daily_components].describe())
#           peak_hr    duration  isi_smooth         dsr
# count  109.000000  109.000000  109.000000  109.000000
# mean    17.293578    2.697248    4.324269    1.553256
# std      1.271502    3.818824    3.220367    1.762854
# min     13.000000    0.000000    0.000000    0.000000
# 25%     17.000000    0.000000    1.234300    0.033800
# 50%     17.000000    0.000000    4.031900    0.991300
# 75%     18.000000    5.000000    6.495300    2.594000
# max     23.000000   14.000000   15.770700    6.948600

# View a distribution of the hour of daily peak burn, which looks like this:
print(report['peak_hr'].value_counts().sort_index())
# peak_hr
# 13     1
# 14     2
# 15     4
# 16    10
# 17    53
# 18    23
# 19    14
# 20     1
# 23     1
# Name: count, dtype: int64

# From here, the outputs can be converted to any datatype for further analysis or
# plotted for visualization.

### Appendix: Timezone ###
# This section will cover how to determine the UTC offset of a Local Standard Time
# (LST) based on a date and location (latitude and longitude). This only applies to
# datasets that are known to be recorded using LST, and is not a substitute in cases
# where a dataset's UTC offset is unknown. See the website FAQ for more details.

# The 'timezonefinder' and 'pytz' packages have functions that can get the timezone
# based on latitude, longitude, and date.
from datetime import datetime
from timezonefinder import TimezoneFinder
from pytz import timezone

# First, make a dataframe of stations with unique ID, latitude, and longitude.
stations = data.loc[:, ['id', 'lat', 'long']].drop_duplicates()
# Print the unique station IDs and coordinates. For this dataset the only station
# is at Petawawa Research Forest (PRF).
print(stations)
#     id     lat    long
# 0  PRF  45.996 -77.427

# Find the timezone based on latitude and longitude.
tf = TimezoneFinder()
tz_loc = tf.timezone_at(lat = stations.at[0, 'lat'], lng = stations.at[0, 'long'])
# Print timezone location. PRF is equivalent to "America/Toronto".
print(tz_loc)
# 'America/Toronto'

# The UTC offset can then be determined from the timezone location and a date. To
# guarantee the UTC offset for LST, use a date in winter (e.g. January 1st in the
# Northern hemisphere or July 1st in the Southern hemisphere).
utc = timezone(tz_loc).localize(datetime(2007, 1, 1)).strftime('%z')
# Print UTC offset, PRF in winter is in Eastern Standard Time.
print(utc)
# '-0500'

# The provided PRF dataset is actually not recorded in LST, but Local Daylight Time
# (Eastern Daylight Time). This is why the timezone parameter is set to -4, and
# why this process is not a substitute for actual information about a dataset.

File: FWI/R/tutorial/tutorial_hourly_FWI.R
#### Hourly Fire Weather Index (FWI) Tutorial ####
# February 2025 (Last updated December 2025)
#
# This script was designed to go with a tutorial on the NG-CFFDRS website to inform
# users how to use scripts associated with FWI2025. Follow along with the
# 'Hourly FWI Tutorial - R' page on the NG-CFFDRS website:
# https://cffdrs.github.io/website_en/tutorials/Hourly_FWI_R
# This tutorial will demonstrate how to generate FWI2025 outputs based on an
# input from a CSV file. The method will differ if using another source file type or
# if integrating code into existing fire management systems. This tutorial
# assumes the user has a working level knowledge of R.
##############################################################################

### Load libraries ###
# Run install.packages() to install any you are missing
library(lubridate)
library(data.table)

### Load functions and data ###
# If the working directory is different from where you saved the FWI2025 scripts,
# change the working directory with setwd() to that folder.

# Load the files containing the variables and functions to calculate FWI2025.
source("NG_FWI.r")
source("util.r")
source("daily_summaries.r")

# Load the input weather station data file
# The path below is specified based on the layout in the GitHub repository
# Change the file path for PRF2007_hourly_wx.csv if your structure is different
data <- read.csv("../../data/PRF2007_hourly_wx.csv")

# Print the column names, data should contain 12 columns
print(names(data))
# [1] "id"       "lat"      "long"     "timezone" "yr"       "mon"
# [7] "day"      "hr"       "temp"     "rh"       "ws"       "prec"

# Previously, the UTC offset (`timezone` column) was a function parameter and could
# be calculated from latitude and longitude. Now it is a data frame column and
# provided. See the appendix of this tutorial for extra information on how to
# calculate the timezone, along with the extra information required about the
# dataset.

### Run FWI2025 ###
# hFWI() is the function that calculates hourly FWI codes in FWI2025. It can
# handle multiple stations and years/fire seasons (not shown in this tutorial). For the arguments you can run `args()`.
args(hFWI)

# For this tutorial, we will leave all the optional parameters to default.
data_fwi <- hFWI(data)
# ########
# Startup values used:
# FFMC = 85.0 or mcffmc = NA %
# DMC = 6.0 and DC = 15.0
# mcgfmc matted = 16.3075 % and standing = 16.3075 %
# cumulative precipitation = 0.0 mm and canopy drying = 0

# Running PRF for 2007
# ########

# Output is a data.frame (matching input class), with FWI calculations appended
# after the input columns. Save the output as a CSV file (overrides any preexisting
# file).
write.csv(data_fwi, "PRF2007_hourly_FWI.csv", row.names = FALSE)

# Print the last two rows of the standard moisture codes and fire behaviour indices.
standard_components <- c("ffmc", "dmc", "dc", "isi", "bui", "fwi")
tail(data_fwi[standard_components], 2)
#         ffmc    dmc       dc    isi    bui    fwi
# 2622 80.2669 3.0924 219.5169 1.6423 5.9745 0.7650
# 2623 82.1759 3.3503 220.0533 2.1462 6.4550 1.2023

# Print a simple summary of the standard FWI components.
summary(data_fwi[standard_components])
#      ffmc            dmc               dc              isi
# Min.   : 0.00   Min.   : 0.000   Min.   : 15.41   Min.   : 0.0000
# 1st Qu.:67.87   1st Qu.: 5.925   1st Qu.: 72.28   1st Qu.: 0.7877
# Median :79.38   Median :16.461   Median :133.06   Median : 1.3958
# Mean   :70.62   Mean   :17.487   Mean   :136.12   Mean   : 2.1582
# 3rd Qu.:85.18   3rd Qu.:26.511   3rd Qu.:188.64   3rd Qu.: 2.9785
# Max.   :94.40   Max.   :50.174   Max.   :292.95   Max.   :17.5958
#      bui              fwi
# Min.   : 0.000   Min.   : 0.000
# 1st Qu.: 8.408   1st Qu.: 0.422
# Median :22.623   Median : 2.408
# Mean   :23.972   Mean   : 3.877
# 3rd Qu.:34.605   3rd Qu.: 5.930
# Max.   :66.148   Max.   :25.135

# Compare your outputs with our standard outputs in PRF2007_standard_hourly_FWI.csv

### Calculate daily summaries ###
# Calculate outputs like peak burn time and number of hours of spread potential.
report <- generate_daily_summaries(data_fwi)

# Print a simple summary of the daily report.
daily_components <- c("peak_hr", "duration", "isi_smooth", "dsr")
summary(report[daily_components])
#    peak_hr         duration        isi_smooth          dsr
# Min.   :13.00   Min.   : 0.000   Min.   : 0.000   Min.   :0.0000
# 1st Qu.:17.00   1st Qu.: 0.000   1st Qu.: 1.234   1st Qu.:0.0338
# Median :17.00   Median : 0.000   Median : 4.032   Median :0.9913
# Mean   :17.29   Mean   : 2.697   Mean   : 4.324   Mean   :1.5533
# 3rd Qu.:18.00   3rd Qu.: 5.000   3rd Qu.: 6.495   3rd Qu.:2.5940
# Max.   :23.00   Max.   :14.000   Max.   :15.771   Max.   :6.9486

# From here, the outputs can be converted to any datatype for further analysis or
# plotted for visualization.

### Appendix: timezone ###
# This section will cover how to calculate a timezone based on a location (latitude
# and longitude) and date. Note that this is not necessarily a substitute for
# actual information about the time used in a dataset. The CFFDRS Weather Guide
# specifies to collect data by local standard time, which is different in many
# places due to the shift to daylight time in summer months (daylight savings).

# The 'lutz' library has functions to get the timezone of the weather station
# based on latitude and longitude.
library(lutz)

# First, make a dataframe of stations with unique ID, latitude, and longitude.
stations <- unique(data[c("id", "lat", "long")])
# Print the unique station IDs and coordinates. For this dataset the only station
# is at Petawawa Research Forest (PRF).
stations
#    id    lat    long
# 1 PRF 45.996 -77.427

# Find the timezone based on latitude and longitude, this can take some time.
# You may need to download the package 'sf' for method = "accurate".
tz_loc <- tz_lookup_coords(stations$lat, stations$long, method = "accurate")
# Print the timezone location. PRF is equivalent to "America/Toronto".
tz_loc
# [1] "America/Toronto"

# The UTC offset can then be determined from the timezone location.
utc <- tz_offset("2007-05-10", tz_loc)[[5]]
# Print the UTC offset
utc

# Since May 10, 2007 is during daylight savings the UTC offset calculated above
# corresponds to Eastern Daylight Time (EDT), UTC-4. This matches the timezone
# column provided since this data was collected using EDT. For Eastern Standard Time
# (EST), the UTC offset would be UTC-5.
